### ISSUE TITLE
Refer backwards?

### ISSUE THREAD

**Issue by anjiro** [pablgonz/scontents#1]
Is it possible to make it so we can retrieve content before it's stored?

```
Soon, I will say, "\getstored{hello}"

Here's some more text.

\Scontents[store-cmd=hello]{Here is some text.}
```

**Comment by PhelypeOleinik** [comment]
I think it's doable, writing things to the `.aux` file...

Though depending on _how much_ stuff you use with `scontents`,
the amount of things written to the `.aux` file can get large pretty
quickly, and cluttering the `.aux` with _all_ that might not be the
greatest idea.

One way to ease that is creating a (say) `.scn` file that contains all
the stuff used with `scontents` (and `\Scontents`), and make the
`.aux` just read that in. This saves space in the `.aux` but still writes
a lot to the `.scn` file, which still takes some time to be read in later.

I think that it might be best to add a key (say) `reuse` which makes
only that one instance of `scontents` (or `\Scontents`) be written to
the external file, greatly reducing the clutter. Of course you can still
say `\setupsc{reuse}` and have a global effect.

What do you (and @pablgonz) think?

**Comment by anjiro** [comment]
I'm attempting a somewhat modular writing style with a grant proposal, where I want have something like:

```
\section{Objectives}
We will accomplish the following:
\begin{itemize}
    \item \getstored{goal:awesome}
    \item \getstored{goal:money}
\end{itemize}

\section{Methods}
\subsection{Awesomeness}
\textit{Goal:} \Scontents[store-cmd=goal:awesome]{We will do super awesome work.} Here
are some details about our breathtaking awesomeness.

\subsection{Spending Money}
\textit{Goal:} \Scontents[store-cmd=goal:money]{We will spend a huge pile of money.} Here's
a list of some nice things we'll buy when we get the grant.
```

So putting things in a separate file doesn't feel good to me. A separate `reuse` key would be okay, especially being global. I don't know how big the `.aux` file would get or how much slowdown we'd be talking about, but my use case doesn't reuse a whole lot of text.

**Comment by PhelypeOleinik** [comment]
What I meant with "writing to the `.aux`" is that, for example here:
```latex
Soon, I will say, "\getstored{hello}"
Here's some more text.
\Scontents[store-cmd=hello, reuse ]{Here is some text.}
```
in the first run `\getstored` does nothing (or prints **hello** or the
classical **??** or whatever), and then `\Scontents[...reuse...]{<stuff>}`
will write `\def\stored@hello{<stuff>}` to the `.aux`. Then, in the
second run, `\getstored` will "know" about `\stored@hello` and will
print that.

Moving stuff _back_ in TeX is not really possible without auxiliary
files...

**Comment by PhelypeOleinik** [comment]
Thinking a bit more about it, there is one issue with implementing this:

`scontents` stores things in an ordered list and you retrieve things by
asking for their number in the sequence, for instance:
```latex
\Scontents[store-cmd=hello, reuse ]{A}
\Scontents[store-cmd=hello, reuse ]{B}
\Scontents[store-cmd=hello]{C} % no reuse here
"\getstored[2]{hello}" and "\getstored[1]{hello}" and "\getstored[3]{hello}"
```
which will typeset: `"B" and "A" and "C"`.

The problem is that the `.aux` file is read at `\begin{document}`,
so if you used `reuse` (perhaps a better name is `remember`) with
those two commands, then in the first run the sequence would
contain `A, B, C` from the three `\Scontents` commands. In the
second run, however, it would contain `A, B, A, B, C` (the first `A, B`
is from the `reuse`d stuff, and the rest is from the `\Scontents`
commands. Then the text typeset would be `"B" and "A" and "A"`,
which is definitely not good...

I think before implementing this we'd need to move to a _named_
approach, rather than numbered...

**Comment by pablgonz** [comment]
@anjiro Good idea :)
@PhelypeOleinik I think it would be feasible, of course, under the cost of having to compile the file twice and write in `.aux` or in a dedicated `.loquesea` file. 

The analysis that you show is more less what I thought, as this we can't get and write in `.aux`, the order of the lists would be lost in the second execution.

I don't really like the idea of _"stressing"_ `.aux` (as tikz does), but, I think it would be a feature to add.

As an idea (only in my mind) I think that the `remember` key (or another name would be the right one), we would have to adapt `\getstored` to execute the reading (starred argument, a key, ...) .

Any ideas in mind?

**Comment by PhelypeOleinik** [comment]
@pablgonz Without compiling twice is not really possible to
move stuff back in TeX.

My concern about cluttering the `.aux` is that arbitrary code
is written there, so if something goes slightly wrong and you
have a bunch of verbatim code there, you may end up with
some pretty awful errors. But this is easily solvable with the
`.loquesea` file ;-)

Furthermore, Plain and ConTeXt don't have a `.aux` file, so a
dedicated `.loquesea` file for `scontents` is the right way.

The biggest problem is the ordering of things... I don't see a
reliable way to implement this feature with the current
item-by-number approach. We'd have to switch to the named
approach before attempting on this, I think...

**Comment by pablgonz** [comment]
> Furthermore, Plain and ConTeXt don't have a `.aux` file, so a
> dedicated `.loquesea` file for `scontents` is the right way.
> 
> The biggest problem is the ordering of things... I don't see a
> reliable way to implement this feature with the current
> item-by-number approach. We'd have to switch to the named
> approach before attempting on this, I think...

Mmmm, this smells to me like `property list`...If I don't remember correctly, it was one of the ideas we discussed at the beginning, maybe we could see how to go down that road and try an implementation that covers these cases.

**Comment by PhelypeOleinik** [comment]
>
> Mmmm, this smells to me like property list...If I don't remember
> correctly, it was one of the ideas we discussed at the beginning, maybe we
> could see how to go down that road and try an implementation that covers
> these cases.
>
Yeah, that was one of the ideas. Basically it's (a lot) easier to
make a named storage store thing by number than a numbered
storage do so by name. I'll see what can be done, eventually.
